src/ConstructiveSolidGeometry/IO.jl

```diff
@@ -1,98 +1,111 @@
-    dictrtop::Dict{Any,Any} = dictr["top"]
-    rtop = _get_r_of_primitive(T, dictrtop, length_unit)
-    rbot, rtop
+volume_primitive_dict = Dict{String, Any}(
+    "tube" => Cone,
+    "cone" => Cone,
+    "sphere" => Sphere,
+    "box" => Box,
+    "HexagonalPrism" => HexagonalPrism,
+    "union" => CSGUnion,
+    "difference" => CSGDifference,
+    "intersection" => CSGIntersection
+)
+
+
+#### INTERNAL PARSE FUNCTIONS
+
+# parses dictionary entries of type Real or String to their value in internal units
+@inline _parse_value(::Type{T}, x::Real, unit::Unitful.Units) where {T} = geom_round(to_internal_units(T(x) * unit))
+@inline _parse_value(::Type{T}, s::String, ::Unitful.Units) where {T} = geom_round(to_internal_units(T(uparse(s))))
+
+# parses dictionary entries of type {"from": ..., "to": ... } to respective AbstractFloat/Interval
+function _parse_interval_from_to(::Type{T}, dict::Union{Dict{String,Any}, Dict{Any,Any}}, unit::Unitful.Units) where {T}
+    To::T = _parse_value(T, dict["to"], unit)
+    From::T = _parse_value(T, dict["from"], unit)
+    (To == -From) ? To : From..To
 end
 
-# converts "r" :  {"from": ... , "to": ...} to the respective AbstractFloat/Interval
-function _get_r_of_primitive(T::DataType, dictr::Union{Dict{String,Any}, Dict{Any,Any}}, length_unit::Unitful.Units, t::Val{:tube} = Val{:tube}())
-    rTo::T = geom_round(T(to_internal_units(internal_length_unit, dictr["to"] * length_unit)))
-    if haskey(dictr, "from")
-        rFrom::T = geom_round(T(to_internal_units(internal_length_unit, dictr["from"] * length_unit)))
-        if rFrom == 0
-            rTo
-        else
-            rFrom..rTo
-        end
-    else
-        rTo
+# parses dictionary entries of type Real, String or {"from": ..., "to": ... } to respective AbstractFloat/Interval
+@inline _parse_interval(::Type{T}, x::Union{Real, String}, unit::Unitful.Units) where {T} = _parse_value(T, x, unit)
+function _parse_interval(::Type{T}, dict::Union{Dict{String,Any}, Dict{Any,Any}}, unit::Unitful.Units) where {T}
+    @assert haskey(dict, "from") && haskey(dict, "to") "Please specify 'from' and 'to' in $(dict)"
+    _parse_interval_from_to(T, dict, unit)
+end
+
+#parses dictionary entry for heights to respective AbstractFloat/Interval (divides by 2 if it is a single value)
+@inline _parse_interval_height(::Type{T}, x::Union{Real, String}, unit::Unitful.Units) where {T} = _parse_value(T, x, unit)/2
+@inline _parse_interval_height(::Type{T}, dict::Union{Dict{String,Any}, Dict{Any,Any}}, unit::Unitful.Units) where {T} = _parse_interval(T, dict, unit)
+
+
+# parses dictionary entry for φ-interval that has {"from" ..., "to": ...} to the respective Nothing/Interval
+function _parse_interval_φ(::Type{T}, dict::Union{Dict{String,Any}, Dict{Any,Any}}, unit::Unitful.Units) where {T}
+    φTo::T = _parse_value(T, dict["to"], unit)
+    φFrom::T = _parse_value(T, dict["from"], unit)
+    if abs(rem2pi(φFrom - φTo, RoundNearest)) > 2*eps(T)
```

Comment:
I would just return the interval.

I have the feeling we can't avoid adjusting the `in` method for `phi`-intervals.

Reply:
I don't understand your comment regarding this specific part of the code.
The problem here was that we need a reliable method to return `φ = nothing` if `2π ` is a full `2π` interval that will skip the `_in_φ` query when calling the `in` method. It is more about how to increase the absolute tolerance which is now set to `2*eps(T)`.

Reply:
This is now solved by not calling `geom_round` in `_parse_value`.